flow/server/flow_subscriber_endpoint.go

```diff
@@ -18,25 +18,72 @@
 package server
 
 import (
+	"github.com/skydive-project/skydive/common"
 	"github.com/skydive-project/skydive/flow"
+	"github.com/skydive-project/skydive/logging"
 	ws "github.com/skydive-project/skydive/websocket"
 )
 
 // FlowSubscriberEndpoint sends all the flows to its subscribers.
 type FlowSubscriberEndpoint struct {
-	pool ws.StructSpeakerPool
+	common.RWMutex
+	pool         ws.StructSpeakerPool
+	nsSubscriber map[string]ws.Speaker
 }
 
+const flowNS = "flow"
+
 // SendFlows sends flow to the subscribers
 func (fs *FlowSubscriberEndpoint) SendFlows(flows []*flow.Flow) {
-	msg := ws.NewStructMessage("flow", "store", flows)
-	fs.pool.BroadcastMessage(msg)
+	fs.RLock()
+	_, ok := fs.nsSubscriber[flowNS]
+	fs.RUnlock()
+
+	if ok {
+		msg := ws.NewStructMessage(flowNS, "store", flows)
+		fs.pool.BroadcastMessage(msg)
+	}
+}
+
+// OnConnected Server interface
+func (fs *FlowSubscriberEndpoint) OnConnected(c ws.Speaker) {
+	ns := flowNS
+
+	if namespaces, ok := c.GetHeaders()["X-Websocket-Namespace"]; ok {
+		if len(namespaces) > 0 {
+			ns = namespaces[0]
+		}
+	}
+
+	fs.Lock()
+	fs.nsSubscriber[ns] = c
+	fs.Unlock()
+
+	logging.GetLogger().Infof("New flow subscriber using namespace: %s", ns)
+}
+
+// OnDisconnected Server interface
+func (fs *FlowSubscriberEndpoint) OnDisconnected(c ws.Speaker) {
+	fs.Lock()
+	defer fs.Unlock()
+
+	for k, nc := range fs.nsSubscriber {
+		if nc == c {
+			delete(fs.nsSubscriber, k)
```

Comment:
if only one namespace per subscriber, the above comment is not valid so we can return here as soon as we delete the entry.